Logging and VTuber exclusivity

MirageAegis opened the pull request:
# What Does This PR Do?
This PR adds configurable moderation logs, VTuber exclusivity, and a few other QoL changes to Sushi Bot.

## Acceptance Criteria Met
- Sushi Bot has `/config logs` for server administrators to configure moderation logs.
- Sushi Bot outputs moderation logs in servers subscribed to logs.
- Sushi Bot leaves every server that doesn't belong to a VTuber.
  - It's checked on start.
  - It can be checked on demand by Sushi Hub administrators.
  - It's checked on new servers that Sushi Bot joins.

## Additional Changes
- Blacklisted users have their reason prepended with "Sushi Bot Blacklist: ".
- Sushi Bot bans all blacklisted users upon joining new eligible servers.
- Sushi Bot logs errors in the error log in Sushi Hub.

## Related Issues
Closes #6 
Closes #24

MirageAegis committed:
#6 Create server schema

Create a schema, model, and wrapper class to represent
a Discord server; which will be used to store server
configurations

MirageAegis committed:
#6 Implement top level config command

Implement the top level /config command for server configurations,
and /config help

MirageAegis committed:
#6 Implement config logs command

Implement the /config logs command for configuring logs channels

MirageAegis committed:
#24 Add server owner verification

MirageAegis committed:
chore: Add error code

MirageAegis committed:
#24 Add admin server singleton

MirageAegis committed:
#24 Add on server join listener

Add a listener for on serber join, and some other events

MirageAegis committed:
#24 Add listener loader

MirageAegis committed:
#24 Load event listeners

MirageAegis committed:
#24 Implement admin reserver command

MirageAegis committed:
build: Upgrade discord.js to v14.12.1

Upgrade discord.js package to get access to newer features
and class properties

MirageAegis committed:
#6 Add colours for embeds

MirageAegis committed:
feat: Add delete method to Server

MirageAegis committed:
#6 Implement embed factories for logging

MirageAegis committed:
#6 Implement event handlers for logging

MirageAegis committed:
#6 Load event handlers

MirageAegis committed:
fix: Log empty messages

Fix bug where Sushi Bot tries to log changes of empty
messages that would result in a crash

MirageAegis committed:
#6 Add message create timestamp

MirageAegis committed:
#6 Use members' display avatars

Use members' display avatars instead of just avatars in
join, leave, ban, and unban events

MirageAegis committed:
#6 Make logs more consistent

MirageAegis committed:
feat: Ban blacklisted users on server join

MirageAegis committed:
chore: Add blacklist statement to blacklist reason

MirageAegis committed:
feat: Ban blacklisted users on server join

MirageAegis committed:
feat: Log generic errors to logs

MirageAegis committed:
feat: Log command errors to logs

MirageAegis committed:
feat: Log command errors to logs

MirageAegis committed:
Merge branch 'logging' of https://github.com/MirageAegis/Sushi-Bot into logging

MirageAegis committed:
feat: Log command errors to logs

MirageAegis committed:
refactor: Remove unused variable

zahatikoff commented on src/commands/configuration/config-subcommands/logs.ts:
```diff
@@ -0,0 +1,82 @@
+
+        await ctx.deferReply();
+
+        // Get the server documetn from the database
```
Typo

MirageAegis replied:
I'll fix that

MirageAegis committed:
refactor: Remove unused import

MirageAegis committed:
docs: Correct code comments

MirageAegis commented on package.json:
```diff
@@ -15,7 +15,7 @@
     "license": "MIT",
     "dependencies": {
-        "discord.js": "^14.11.0",
+        "discord.js": "^14.12.1",
```
**IMPORTANT!!**
The discord.js dependency has been updated from v14.11.0 to v14.12.1
Make sure to run `npm ci` to make a clean install of the updated dependency

MirageAegis commented on src/commands/configuration/config.ts:
```diff
@@ -0,0 +1,91 @@
+    const subcmdPath: string = path.join(subcmdFolder, file);
+    const subcmd: Subcommand = require(subcmdPath).command;
+
+    const name: string = subcmd.data.name;
+
+    subcommands.set(name, subcmd);
+    data.addSubcommand(subcmd.data);
+}
+
+export const command: Command = {
+    // Command headers
+    data: data,
+
+    // Command execution
+    async execute(ctx: ChatInputCommandInteraction): Promise<void> {
+        const subcmd = ctx.options.getSubcommand();
+
+        // Execute the requested subcommand
+        await subcommands.get(subcmd).execute(ctx);
+    },
+
+    // Error handler
+    async error(ctx: ChatInputCommandInteraction, err: Error): Promise<void> {
+        const subcmd = ctx.options.getSubcommand();
+
+        // Execute the requested subcommand's error handler
+        await subcommands.get(subcmd).error(ctx, err);
+    },
+
+    // Help command embed
+    help: new EmbedBuilder()
+        .setTitle("Config")
+        .setDescription(
+            `Commands for configuring events in a server. Use \`/${name} help\` for information about these commands. Can only be used by server administrators`
+        )
+        .addFields(
+            { name: "Format", value: `\`/${name} <subcommand>\`` },
+            { name: "<subcommand>", value: "The subcommand to use" }
+        )
+};
```
This is the top level `/config` command used for server configurations

MirageAegis commented on src/commands/configuration/config-subcommands/help.ts:
```diff
@@ -0,0 +1,138 @@
+const choices = Array.from(commands.keys()).map((cmd: string): { name: string, value: string } => {
+    return { name: cmd, value: cmd };
+});
+
+export const command: Subcommand = {
+    // Command headers
+    data: new SlashCommandSubcommandBuilder()
+        .setName(name)
+        .setDescription("Need help with my server configuration commands?")
+        .addStringOption(o =>
+            o.setName("command")
+                .setDescription("The command to get more information about")
+                .addChoices(...choices)
+        ),
+
+    // Command execution
+    async execute(ctx: ChatInputCommandInteraction): Promise<void> {
+        // Get the topic if it exists
+        const command = ctx.options.getString("command") ?? null;
+        
+        // If there is a topic, send its embed
+        if (command) {
+            // Set the embed author field to "Help" and add the server's icon
+            const embed = commands.get(command)
+                .setAuthor({ name: "Help", iconURL: ctx.guild.iconURL() });
+            await ctx.reply({ embeds: [embed] });
+            return;
+        } else { // Otherwise send the default help window
+            defaultEmbed.setAuthor({ name: "Help", iconURL: ctx.guild.iconURL() });
+            await ctx.reply({ embeds: [defaultEmbed] });
+            return;
+        }
+    },
+
+    // Error handler
+    error: defaultErrorHandler,
+
+    // Help command embed
+    help: help
+};
```
This is the `/config help` which works similarly to `/help` and `/admin help`

MirageAegis commented on src/commands/configuration/config-subcommands/logs.ts:
```diff
@@ -0,0 +1,82 @@
+            { name: "Format", value: `\`/config ${name} [logs]\`` },
+            { name: "[logs]", value: "Optional parameter. The channel to set as logging channel. Omit to remove the logs channel" }
+        )
+};
```
This is the `/config logs` command which is used to set up moderation logs
It saves the logs channel's channel ID to the database

MirageAegis commented on src/events/events.ts:
```diff
@@ -0,0 +1,391 @@
+} from "./logs";
+import { Blacklist } from "../schemas/blacklist";
+
+/*
+ * This module has event listeners for the bot
+ */
```
This module has the event listeners used for moderation logs and VTuber exclusivity

MirageAegis commented on src/events/events.ts:
```diff
@@ -0,0 +1,391 @@
+        // Log to the logs channel
+        await logs.send({ embeds: [genMemberBanEmbed(ban)] });
+    } catch {
+        // Remove configurations if logging failed.
+        // Most likely causes are that the channel was deleted or that
+        // Sushi Bot had its permissions revoked
+        server.logs = null;
+        await server.save();
+    }
+};
+
+/**
+ * Logs member unban events to servers subscribed to logs.
+ * 
+ * @param client the Discord bot
+ * @param ban the ban data
+ */
+export const onMemberUnban = async (client: Client, ban: GuildBan): Promise<void> => {
+    // The server and logs channel ID of the server that the member was unbanned from
+    const server: Server = await Server.get(ban.guild.id);
+    const logsID: string = server.logs;
+
+    // Skip if the server isn't subscribed to logs
+    if (!logsID) {
+        return;
+    }
+
+    const logs: TextChannel = <TextChannel> client.channels.cache.get(logsID);
+
+    try {
+        // Log to the logs channel
+        await logs.send({ embeds: [genMemberUnbanEmbed(ban)] });
+    } catch {
+        // Remove configurations if logging failed.
+        // Most likely causes are that the channel was deleted or that
+        // Sushi Bot had its permissions revoked
+        server.logs = null;
+        await server.save();
+    }
+};
```
These four listeners are quite similar. They get triggered when a server gains or loses a member and logs it to the moderation logs of servers subscribed to logs

MirageAegis commented on src/events/logs.ts:
```diff
@@ -0,0 +1,385 @@
+    const embed: EmbedBuilder = new EmbedBuilder()
+        .setTitle(ban.user.username)
+        .setColor(RED)
+        .setAuthor({
+            name: "Member banned",
+            iconURL: ban.guild.iconURL()
+        })
+        .setThumbnail(ban.user.displayAvatarURL())
+        .addFields(
+            { name: "ID", value: ban.user.id, inline: true },
+            { name: "Created", value: `<t:${Math.floor(ban.user.createdTimestamp / millisToSecs)}>`, inline: true },
+            { name: "Bot user", value: `${ban.user.bot}`, inline: true }
+        );
+
+    return embed;
+};
+
+/**
+ * Generates an embed for whenever a user is unbanned from a server.
+ * 
+ * @param ban the ban removed
+ * @returns an embed to be logged
+ */
+export const genMemberUnbanEmbed = (ban: GuildBan): EmbedBuilder => {
+    const embed: EmbedBuilder = new EmbedBuilder()
+        .setTitle(ban.user.username)
+        .setColor(BLUE)
+        .setAuthor({
+            name: "Member banned",
+            iconURL: ban.guild.iconURL()
+        })
+        .setThumbnail(ban.user.displayAvatarURL())
+        .addFields(
+            { name: "ID", value: ban.user.id, inline: true },
+            { name: "Created", value: `<t:${Math.floor(ban.user.createdTimestamp / millisToSecs)}>`, inline: true },
+            { name: "Bot user", value: `${ban.user.bot}`, inline: true }
+        );
+
+    return embed;
+};
```
These are the embeds used by the listeners above

MirageAegis commented on src/events/events.ts:
```diff
@@ -0,0 +1,391 @@
+    }
+};
+
+/**
+ * Logs member server profile update events to servers subscribed to logs.
+ * 
+ * @param client the Discord bot
+ * @param before the member's profile state before update
+ * @param after the member's current profile state
+ */
+export const onMemberUpdate = async (client: Client, before: GuildMember, after: GuildMember): Promise<void> => {
+    // The server and logs channel ID of the server that the member updated their profile in
+    const server: Server = await Server.get(before.guild.id);
+    const logsID: string = server.logs;
+
+    // Skip if the server isn't subscribed to logs
+    if (!logsID) {
+        return;
+    }
+
+    const logs: TextChannel = <TextChannel> client.channels.cache.get(logsID);
+    const embed: EmbedBuilder = genMemberUpdateEmbed(before, after);
+
+    // If embed is null, the changes aren't significant to Sushi Bot
+    if (!embed) {
+        return;
+    }
+
+    try {
+        // Log to the logs channel
+        await logs.send({ embeds: [embed] });
+    } catch {
+        // Remove configurations if logging failed.
+        // Most likely causes are that the channel was deleted or that
+        // Sushi Bot had its permissions revoked
+        server.logs = null;
+        await server.save();
+    }
+};
```
This listener gets triggered when a server member updates their server profile

MirageAegis committed:
refactor: Remove stray column in on member update embed

MirageAegis commented on src/events/logs.ts:
```diff
@@ -0,0 +1,385 @@
+    }
+
+    return embed;
+};
```
This is the embed used in the listener above
It checks for nickname and server avatar updates

zahatikoff commented on src/commands/admin/admin-subcommands/reserver.ts:
**Reserver** does not seem to be the most fitting name for the command, as it mostly just reloads the bot?

MirageAegis replied:
It only reloads the servers that the bot is in but I can see why it doesn't seem very fitting. We can change it in the future if we come up with a better name, or we can merge the command with `/admin reblacklist` and call it `/admin reload`

zahatikoff replied:
Yeah also sorry for a possible shitload of notiffs
I got a lil confused with the code review tool

MirageAegis replied:
No worries!
I'll get on with the `/admin reload` command when I have time. In the meantime, you can go through the comments I made and resolve them as you read them

MirageAegis commented on src/events/events.ts:
```diff
@@ -0,0 +1,391 @@
+    const guilds: Collection<string, Guild> = client.guilds.cache;
+
+    // For each server...
+    for await (const [, guild] of guilds) {
+        try {
+            // If fetched, the user is a member of that server
+            await guild.members.fetch(before.id);
+        } catch {
+            // If not, they're not, so we skip
+            continue;
+        }
+
+        // The server and logs channel ID of the mutual server
+        const server: Server = await Server.get(guild.id);
+        const logsID: string = server.logs;
+
+        // Skip if the server isn't subscribed to logs
+        if (!logsID) {
+            continue;
+        }
+
+        const logs: TextChannel = <TextChannel> client.channels.cache.get(logsID);
+
+        try {
+            // Log to the logs channel
+            await logs.send({ embeds: [
+                embed.setAuthor({
+                    name: "Profile updated",
+                    iconURL: guild.iconURL()
+                })
+            ]});
+        } catch {
+            // Remove configurations if logging failed.
+            // Most likely causes are that the channel was deleted or that
+            // Sushi Bot had its permissions revoked
+            server.logs = null;
+            await server.save();
+        }
+    }
+};
```
This listener gets triggered when a user makes changes to their global profile; and broadcasts the changes, if they are significant to Sushi Bot's logs, to all servers' logs channels

MirageAegis commented on src/events/logs.ts:
```diff
@@ -0,0 +1,385 @@
+        oldDisplayName === newDisplayName &&
+        oldPfp === newPfp
+    ) {
+        return null;
+    }
+
+    const embed: EmbedBuilder = new EmbedBuilder()
+        .setTitle(oldUsername)
+        .setColor(PURPLE)
+        .setThumbnail(before.displayAvatarURL())
+        .addFields(
+            { name: "ID", value: before.id }
+        );
+
+    // Add this if the username was updated
+    if (oldUsername !== newUsername) {
+        embed.addFields(
+            { name: "Old username", value: oldUsername, inline: false },
+            { name: "New username", value: newUsername, inline: false },
+        );
+    }
+
+    // Add this if the display name was updated
+    if (oldDisplayName !== newDisplayName) {
+        embed.addFields(
+            { name: "Old display name", value: oldDisplayName, inline: false },
+            { name: "New display name", value: newDisplayName, inline: false },
+        );
+    }
+
+    // Add this if the username was updated
+    if (oldPfp !== newPfp) {
+        embed.addFields(
+            { name: "New avatar", value: "_ _", inline: false }
+        )
+            .setImage(after.displayAvatarURL());
+    }
+
+    return embed;
+};
```
This is the embed used for the listener above
It checks for changes to usernames, display names, and avatars

MirageAegis commented on src/events/events.ts:
```diff
@@ -0,0 +1,391 @@
+    // If embed is null, the message wasn't significant to Sushi Bot
+    if (!embed) {
+        return;
+    }
+
+    const logs: TextChannel = <TextChannel> client.channels.cache.get(logsID);
+
+    try {
+        // Log to the logs channel
+        await logs.send({ embeds: [embed] });
+    } catch {
+        // Remove configurations if logging failed.
+        // Most likely causes are that the channel was deleted or that
+        // Sushi Bot had its permissions revoked
+        server.logs = null;
+        await server.save();
+    }
+};
+
+/**
+ * Triggered when the bot joins a Discord server.
+ * It leaves the server if the owner is ineligible to have the bot.
+ * 
+ * @param client the Discord bot
+ * @param server the Discord server joined
+ */
+export const onServerJoin = async (client: Client, server: Guild): Promise<void> => {
+    const eligible: boolean = await leaveIneligibleServer(client, server, getAdminLogsChannel());
+
+    if (eligible) {
+        const bl: Blacklist = await Blacklist.get();
+
+        for await (const [uid, reason] of bl.users) {
+            await server.bans.create(
+                uid,
+                { reason: reason }
+            );
+        }
+    }
+};
```
These two get triggered when the state of a message changes

MirageAegis commented on src/events/logs.ts:
```diff
@@ -0,0 +1,385 @@
+    if (!content && !attachments.size) {
+        return null;
+    }
+
+    const embed: EmbedBuilder = new EmbedBuilder()
+        .setTitle(author.username)
+        .setColor(ORANGE)
+        .setAuthor({
+            name: "Message deleted",
+            iconURL: message.guild.iconURL()
+        })
+        .setTimestamp(message.createdAt)
+        .addFields(
+            { name: "Member ID", value: author.id, inline: true },
+            { name: "Message ID", value: message.id, inline: true },
+            { name: "Channel", value: message.channel.toString(), inline: false },
+        );
+
+    // Add if the message has text
+    if (content) {
+        embed.addFields(
+            { name: "Content", value: content, inline: false }
+        );
+    }
+
+    // Add if the message has attachments
+    if (attachments.size) {
+        let urls: string = "";
+
+        for (const [, attachment] of attachments) {
+            urls += `${attachment.url}\n`;
+        }
+
+        embed.addFields(
+            { name: "Attachments", value: urls, inline: false }
+        );
+    }
+
+    return embed;
+};
```
These are the embeds used for message changes
They check the message's content and attachments

MirageAegis commented on src/schemas/server.ts:
```diff
@@ -0,0 +1,154 @@
+
+import mongoose, { Schema, Document, InferSchemaType } from "mongoose";
+
+const serverSchema: Schema = new Schema({
+    _id: String,
+    logs: String,
+});
+
+type ServerT = InferSchemaType<typeof serverSchema>;
```
This is the database schema used to represent server configurations, more fields will be added in the future

MirageAegis commented on src/schemas/server.ts:
```diff
@@ -0,0 +1,154 @@
+    public set logs(id: string) {
+        (<ServerT> this.data).logs = id;
+    }
+}
```
This is the wrapper class for the database schema, it has all the necessary functions and methods

MirageAegis commented on src/schemas/server.ts:
```diff
@@ -0,0 +1,154 @@
+     */
+    private static readonly model = mongoose.model("Server", serverSchema);
+
+    /**
+     * The cached servers.
+     * The servers are stored along with the ID for the timer responsible for
+     * clearing the memory.
+     */
+    private static cache: Map<string, [Server, NodeJS.Timeout]> = new Map();
```
The server configurations that are fetched from the database are cached to improve performance at the cost of memory

MirageAegis commented on src/schemas/server.ts:
```diff
@@ -0,0 +1,154 @@
+        }
+    }
+
+    /**
+     * Gets a server document from the database.
+     * 
+     * @param id the server ID
+     * @returns the requested server document or null if none was found
+     */
+    public static async get(id: string): Promise<Server> {
+        // Try to get the server from the cache
+        const s: [Server, NodeJS.Timeout] = Server.cache.get(id);
+        let server: Server;
+
+        if (s) {
+            // If the server is found, clear the old timeout
+            clearTimeout(s[1]);
+            server = s[0];
+        } else {
+            // Otherwise, fetch from the database
+            const data: Document = await Server.model.findById(id);
+            
+            // If the database didn't have the server,
+            // create a new document
+            server = data ? new Server(data) : new Server(id);
+        }
+
+        // Create a timeout that flushes the server from the cache
+        // after a set period of time
+        const timeout: NodeJS.Timeout = setTimeout(async (): Promise<void> => {
+            await server.save();
+            Server.cache.delete(id);
+        }, serverLifetime);
+
+        // Save to the cache
+        Server.cache.set(id, [server, timeout]);
+
+        return server;
+    }
```
They are cached and refreshed in the `Server.get` function

MirageAegis committed:
fix: Clear deleted server configs from cache

MirageAegis commented on src/events/events.ts:
```diff
@@ -0,0 +1,391 @@
+    }
+};
+
+/**
+ * Triggered when the bot leaves a Discord server,
+ * or when a Discord server is deleted.
+ * It deletes the server configurations from the database.
+ * 
+ * @param client the Discord bot
+ * @param server the Discord server left
+ */
+export const onServerLeave = async (client: Client, server: Guild): Promise<void> => {
+    await (await Server.get(server.id)).delete();
+};
```
This gets triggered when Sushi Bot leaves a server
It deletes the server configurations for that server

MirageAegis commented on src/events/events.ts:
````diff
@@ -0,0 +1,391 @@
+    await (await Server.get(server.id)).delete();
+};
+
+/**
+ * Triggered when a non-command error occurs.
+ * Errors are logged to the admin logs in the admin server.
+ * 
+ * @param client the Discord bot
+ * @param error the error
+ */
+export const onError = async (client: Client, error: Error): Promise<void> => {
+    const logs: TextChannel = getAdminLogsChannel();
+
+    const report: string = "```\n" +
+                           "Generic Error\n\n" +
+                           `${error}\n` +
+                           "```";
+    
+    await logs.send(report);
+};
````
This is the error handler for generic errors
It logs the errors to the error log in the admin server

MirageAegis commented on src/util/error-handler.ts:
````diff
@@ -74,4 +80,23 @@ export const defaultErrorHandler: ErrorHandler = async (ctx: ChatInputCommandInt
+
+    // ----- ERROR LOG -----
+
+    const user: User = ctx.user;
+    const command: ApplicationCommand = ctx.command;
+
+    const report: string = "```\n" +
+                           "Command Error\n\n" +
+                           `Server: ${ctx.guild.name}\n` +
+                           `User: ${user.id} (${user.username})\n` +
+                           `Command: ${command.name}\n` +
+                           `Options: ${command.options.toString()}\n\n` +
+                           // Error name, and error code if it's a Discord API error
+                           `${err.name}${err instanceof DiscordAPIError ? `: ${err.code} (${err.message})` : ""}\n` +
+                           "```";
+    
+    await getAdminLogsChannel().send(report);
````
This has been added to the command error handler to make it log to the error log in the admin server

MirageAegis commented on src/util/init.ts:
```diff
@@ -54,8 +72,54 @@ export const init = async (client: Client): Promise<void> => {
+    client.on(Events.GuildCreate, async (server: Guild): Promise<void> => {
+        await onServerJoin(client, server);
+    });
+
+    client.on(Events.GuildMemberAdd, async (member: GuildMember): Promise<void> => {
+        await onMemberJoin(client, member);
+    });
+
+    client.on(Events.GuildMemberRemove, async (member: GuildMember): Promise<void> => {
+        await onMemberLeave(client, member);
+    });
+
+    client.on(Events.GuildBanAdd, async (ban: GuildBan): Promise<void> => {
+        await onMemberBan(client, ban);
+    });
+
+    client.on(Events.GuildBanRemove, async (ban: GuildBan): Promise<void> => {
+        await onMemberUnban(client, ban);
+    });
+
+    client.on(Events.GuildMemberUpdate, async (before: GuildMember, after: GuildMember): Promise<void> => {
+        await onMemberUpdate(client, before, after);
+    });
+
+    client.on(Events.UserUpdate, async (before: User, after: User): Promise<void> => {
+        await onUserUpdate(client, before, after);
+    });
+
+    client.on(Events.MessageUpdate, async (before: Message, after: Message): Promise<void> => {
+        await onMessageEdit(client, before, after);
+    });
+
+    client.on(Events.MessageDelete, async (message: Message): Promise<void> => {
+        await onMessageDelete(client, message);
+    });
+
+    client.on(Events.Error, async (error: Error): Promise<void> => {
+        await onError(client, error);
+    });
+};
```
This is where all the listeners are loaded into the bot

MirageAegis reviewed (commented):
This review will walk through key points of this PR, which mainly consists of implementations for #6 and #24

MirageAegis committed:
docs: Fix typo

zahatikoff reviewed (approved):
I don't think the `/admin reserver` command name is fitting, but since it's internal use only it's fine for now

MirageAegis committed:
refactor: Merge reload commands

Merge the admin reblacklist and admin reserver commands
into admin reload, which triggers both
